plutus-chain-index-core/src/Plutus/ChainIndex/Handlers.hs

```diff
@@ -273,6 +278,47 @@ getTxOutSetAtAddress pageQuery cred = do
       pure $ QueryResponse txouts (nextPageQuery page)
 
 
+getDatumsAtAddress ::
+  forall effs.
+    ( Member (State ChainIndexState) effs
+    , Member (BeamEffect Sqlite) effs
+    , Member (LogMsg ChainIndexLog) effs
+    )
+  => PageQuery TxOutRef
+  -> Credential
+  -> Eff effs (QueryResponse [Datum])
+getDatumsAtAddress pageQuery (toDbValue -> cred) = do
+  utxoState <- gets @ChainIndexState UtxoState.utxoState
+  case UtxoState.tip utxoState of
+    TipAtGenesis -> do
+      logWarn TipIsGenesis
+      pure (QueryResponse [] Nothing)
+
+    _             -> do
+      let emptyHash = toDbValue $ DatumHash emptyByteString
+          queryPage =
+            fmap _addressRowOutRef
+            $ filter_ (\row ->
+                         ( _addressRowCred row ==. val_ cred )
+                         &&. (_addressRowDatumHash row /=. val_ emptyHash) )
+            $ all_ (addressRows db)
+          queryAll =
+            select
+            $ filter_ (\row -> _addressRowCred row ==. val_ cred
+                       &&. (_addressRowDatumHash row /=. val_ emptyHash ) )
+            $ all_ (addressRows db)
+      pRefs <- selectPage (fmap toDbValue pageQuery) queryPage
+      let page = fmap fromDbValue pRefs
+      row_l <- List.filter (\(_, t, _) -> List.elem t (pageItems page)) <$> queryList queryAll
```

Comment:
This is a nasty workaround for a limitation in `selectPage`.

It's type is now:

```haskell
PageQuery a
    -> Q dbt db BeamThreadingArg (QExpr dbt BeamThreadingArg a)
    -> Eff effs (Page a)
```

but it could be

```haskell
PageQuery index
    -> Q dbt db BeamThreadingArg (QExpr dbt BeamThreadingArg row)
    -> (row -> index)
    -> Eff effs (Page row)
```

I.e. making it possible to have rows that contain more than the index you're using for pagination, by passing a function to extract the index from the row.

You call if you want to fix that here or to have this be a ticket for later.

Reply:
I'll create a ticket for it.